lib/elixir/lib/macro.ex

```diff
@@ -131,7 +131,55 @@ defmodule Macro do
 
   @typep output_expr :: {output_expr | atom, metadata, atom | [output]}
 
-  @typedoc "A keyword list of AST metadata"
+  @typedoc """
+  A keyword list of AST metadata.
+
+  The metadata in Elixir AST is a keyword list of values. Any key can be used
+  and different parts of the compiler may use different keys. For example,
+  the AST received by a macro will always include the `:line` annotation,
+  while the AST emitted by a quote never has the `:line` annotation.
+
+  The following metadata is public:
+
+    * `:column` - The column number of the AST node. It is included only
+      in AST generated by `Code.string_to_quoted/2` with the `columns: true`
+      option
+    * `:context` - Defines the context in which the AST was generated.
+      For example, `quote/2` will include the module calling `quote/2`
+      as the context. This often used to distinguish code from code generated
+      by a macro/quote
+    * `:counter` - The variable counter used for variable hygiene. In terms of
+      AST, each variable is identified by a `{name, metadata[:counter] || context}`
```

Comment:
I would clarify that `{name, metadata[:counter] || context}` is not the AST representation of a variable.

Reply:
@fertapric How about `In terms of AST` => `In terms of the compiler`

Technically AST doesn't really identify a variable anyways, it's the compiler that decides about the identity of the variable.

We could also just write that a variable is identified by either name & counter or name & context and leave out the tuple.

Reply:
> We could also just write that a variable is identified by either name & counter or name & context and leave out the tuple.

I like this one 👌

Reply:
```md
* `:counter` - The variable counter used for variable hygiene. In terms of
  the compiler, each variable is identified by the combination of either
  `name` & `metadata[:counter]` or `name` & `context`.
```